synapse/storage/databases/main/event_federation.py

```diff
@@ -735,34 +744,69 @@ async def get_oldest_event_ids_with_depth_in_room(
-            # some other event referenced them by prev_event and aren't peristed
-            # in our database yet (meaning we don't know their depth
+            # some other event referenced them by prev_event and aren't
+            # persisted in our database yet (meaning we don't know their depth
             # specifically). So we need to look for the aproximate depth from
             # the events connected to the current backwards extremeties.
             sql = """
-                SELECT b.event_id, MAX(e.depth) FROM events as e
+                SELECT backward_extrem.event_id, MAX(event.depth) FROM events as event
                 /**
                  * Get the edge connections from the event_edges table
                  * so we can see whether this event's prev_events points
                  * to a backward extremity in the next join.
                  */
-                INNER JOIN event_edges as g
-                ON g.event_id = e.event_id
+                INNER JOIN event_edges as edge
+                ON edge.event_id = event.event_id
                 /**
                  * We find the "oldest" events in the room by looking for
                  * events connected to backwards extremeties (oldest events
                  * in the room that we know of so far).
                  */
-                INNER JOIN event_backward_extremities as b
-                ON g.prev_event_id = b.event_id
-                WHERE b.room_id = ? AND g.is_state is ?
-                GROUP BY b.event_id
+                INNER JOIN event_backward_extremities as backward_extrem
+                ON edge.prev_event_id = backward_extrem.event_id
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                INNER JOIN event_backfill_attempts as backfill_attempt_info
+                ON backfill_attempt_info.event_id = backward_extrem.event_id
+                WHERE
+                    backward_extrem.room_id = ?
+                    /* We only care about normal events because TODO: why? */
+                    AND edge.is_state is ? /* False */
+                    /**
+                     * We only want backwards extremities that are older than or at
+                     * the same position of the given `current_depth` (where older
+                     * means less than a given depth).
+                     */
+                    AND MAX(event.depth) <= current_depth
+                    /**
+                     * Exponential back-off (up to the upper bound) so we don't retry the 
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc
+                     */
+                    AND ? /* current_time */ >= backfill_attempt_info.last_attempt_ts + least(2^backfill_attempt_info.num_attempts * ?, ? /* upper bound */)
```

Comment:
Is implementing the exponential back-off lookup in the SQL query acceptable?

I could imagine changing the schema `event_backfill_attempts` to include the time it should retry at but then that bakes in the time to the database and the config could change.

Reply:
Commented on https://github.com/matrix-org/synapse/pull/13589#discussion_r955588328 to make sure we're going down the right path

Reply:
Based on no apprehension to include that info in the database table from https://github.com/matrix-org/synapse/pull/13589, I think we're happy to do this ⏩